<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en">
<head>
<!-- 2021-06-10 Thu 11:24 -->
<meta http-equiv="Content-Type" content="text/html;charset=utf-8" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<title>Discordant variant analysis</title>
<meta name="generator" content="Org mode" />
<meta name="author" content="Daniel Coral" />
<style type="text/css">
 <!--/*--><![CDATA[/*><!--*/
  .title  { text-align: center;
             margin-bottom: .2em; }
  .subtitle { text-align: center;
              font-size: medium;
              font-weight: bold;
              margin-top:0; }
  .todo   { font-family: monospace; color: red; }
  .done   { font-family: monospace; color: green; }
  .priority { font-family: monospace; color: orange; }
  .tag    { background-color: #eee; font-family: monospace;
            padding: 2px; font-size: 80%; font-weight: normal; }
  .timestamp { color: #bebebe; }
  .timestamp-kwd { color: #5f9ea0; }
  .org-right  { margin-left: auto; margin-right: 0px;  text-align: right; }
  .org-left   { margin-left: 0px;  margin-right: auto; text-align: left; }
  .org-center { margin-left: auto; margin-right: auto; text-align: center; }
  .underline { text-decoration: underline; }
  #postamble p, #preamble p { font-size: 90%; margin: .2em; }
  p.verse { margin-left: 3%; }
  pre {
    border: 1px solid #ccc;
    box-shadow: 3px 3px 3px #eee;
    padding: 8pt;
    font-family: monospace;
    overflow: auto;
    margin: 1.2em;
  }
  pre.src {
    position: relative;
    overflow: visible;
    padding-top: 1.2em;
  }
  pre.src:before {
    display: none;
    position: absolute;
    background-color: white;
    top: -10px;
    right: 10px;
    padding: 3px;
    border: 1px solid black;
  }
  pre.src:hover:before { display: inline;}
  /* Languages per Org manual */
  pre.src-asymptote:before { content: 'Asymptote'; }
  pre.src-awk:before { content: 'Awk'; }
  pre.src-C:before { content: 'C'; }
  /* pre.src-C++ doesn't work in CSS */
  pre.src-clojure:before { content: 'Clojure'; }
  pre.src-css:before { content: 'CSS'; }
  pre.src-D:before { content: 'D'; }
  pre.src-ditaa:before { content: 'ditaa'; }
  pre.src-dot:before { content: 'Graphviz'; }
  pre.src-calc:before { content: 'Emacs Calc'; }
  pre.src-emacs-lisp:before { content: 'Emacs Lisp'; }
  pre.src-fortran:before { content: 'Fortran'; }
  pre.src-gnuplot:before { content: 'gnuplot'; }
  pre.src-haskell:before { content: 'Haskell'; }
  pre.src-hledger:before { content: 'hledger'; }
  pre.src-java:before { content: 'Java'; }
  pre.src-js:before { content: 'Javascript'; }
  pre.src-latex:before { content: 'LaTeX'; }
  pre.src-ledger:before { content: 'Ledger'; }
  pre.src-lisp:before { content: 'Lisp'; }
  pre.src-lilypond:before { content: 'Lilypond'; }
  pre.src-lua:before { content: 'Lua'; }
  pre.src-matlab:before { content: 'MATLAB'; }
  pre.src-mscgen:before { content: 'Mscgen'; }
  pre.src-ocaml:before { content: 'Objective Caml'; }
  pre.src-octave:before { content: 'Octave'; }
  pre.src-org:before { content: 'Org mode'; }
  pre.src-oz:before { content: 'OZ'; }
  pre.src-plantuml:before { content: 'Plantuml'; }
  pre.src-processing:before { content: 'Processing.js'; }
  pre.src-python:before { content: 'Python'; }
  pre.src-R:before { content: 'R'; }
  pre.src-ruby:before { content: 'Ruby'; }
  pre.src-sass:before { content: 'Sass'; }
  pre.src-scheme:before { content: 'Scheme'; }
  pre.src-screen:before { content: 'Gnu Screen'; }
  pre.src-sed:before { content: 'Sed'; }
  pre.src-sh:before { content: 'shell'; }
  pre.src-sql:before { content: 'SQL'; }
  pre.src-sqlite:before { content: 'SQLite'; }
  /* additional languages in org.el's org-babel-load-languages alist */
  pre.src-forth:before { content: 'Forth'; }
  pre.src-io:before { content: 'IO'; }
  pre.src-J:before { content: 'J'; }
  pre.src-makefile:before { content: 'Makefile'; }
  pre.src-maxima:before { content: 'Maxima'; }
  pre.src-perl:before { content: 'Perl'; }
  pre.src-picolisp:before { content: 'Pico Lisp'; }
  pre.src-scala:before { content: 'Scala'; }
  pre.src-shell:before { content: 'Shell Script'; }
  pre.src-ebnf2ps:before { content: 'ebfn2ps'; }
  /* additional language identifiers per "defun org-babel-execute"
       in ob-*.el */
  pre.src-cpp:before  { content: 'C++'; }
  pre.src-abc:before  { content: 'ABC'; }
  pre.src-coq:before  { content: 'Coq'; }
  pre.src-groovy:before  { content: 'Groovy'; }
  /* additional language identifiers from org-babel-shell-names in
     ob-shell.el: ob-shell is the only babel language using a lambda to put
     the execution function name together. */
  pre.src-bash:before  { content: 'bash'; }
  pre.src-csh:before  { content: 'csh'; }
  pre.src-ash:before  { content: 'ash'; }
  pre.src-dash:before  { content: 'dash'; }
  pre.src-ksh:before  { content: 'ksh'; }
  pre.src-mksh:before  { content: 'mksh'; }
  pre.src-posh:before  { content: 'posh'; }
  /* Additional Emacs modes also supported by the LaTeX listings package */
  pre.src-ada:before { content: 'Ada'; }
  pre.src-asm:before { content: 'Assembler'; }
  pre.src-caml:before { content: 'Caml'; }
  pre.src-delphi:before { content: 'Delphi'; }
  pre.src-html:before { content: 'HTML'; }
  pre.src-idl:before { content: 'IDL'; }
  pre.src-mercury:before { content: 'Mercury'; }
  pre.src-metapost:before { content: 'MetaPost'; }
  pre.src-modula-2:before { content: 'Modula-2'; }
  pre.src-pascal:before { content: 'Pascal'; }
  pre.src-ps:before { content: 'PostScript'; }
  pre.src-prolog:before { content: 'Prolog'; }
  pre.src-simula:before { content: 'Simula'; }
  pre.src-tcl:before { content: 'tcl'; }
  pre.src-tex:before { content: 'TeX'; }
  pre.src-plain-tex:before { content: 'Plain TeX'; }
  pre.src-verilog:before { content: 'Verilog'; }
  pre.src-vhdl:before { content: 'VHDL'; }
  pre.src-xml:before { content: 'XML'; }
  pre.src-nxml:before { content: 'XML'; }
  /* add a generic configuration mode; LaTeX export needs an additional
     (add-to-list 'org-latex-listings-langs '(conf " ")) in .emacs */
  pre.src-conf:before { content: 'Configuration File'; }

  table { border-collapse:collapse; }
  caption.t-above { caption-side: top; }
  caption.t-bottom { caption-side: bottom; }
  td, th { vertical-align:top;  }
  th.org-right  { text-align: center;  }
  th.org-left   { text-align: center;   }
  th.org-center { text-align: center; }
  td.org-right  { text-align: right;  }
  td.org-left   { text-align: left;   }
  td.org-center { text-align: center; }
  dt { font-weight: bold; }
  .footpara { display: inline; }
  .footdef  { margin-bottom: 1em; }
  .figure { padding: 1em; }
  .figure p { text-align: center; }
  .equation-container {
    display: table;
    text-align: center;
    width: 100%;
  }
  .equation {
    vertical-align: middle;
  }
  .equation-label {
    display: table-cell;
    text-align: right;
    vertical-align: middle;
  }
  .inlinetask {
    padding: 10px;
    border: 2px solid gray;
    margin: 10px;
    background: #ffffcc;
  }
  #org-div-home-and-up
   { text-align: right; font-size: 70%; white-space: nowrap; }
  textarea { overflow-x: auto; }
  .linenr { font-size: smaller }
  .code-highlighted { background-color: #ffff00; }
  .org-info-js_info-navigation { border-style: none; }
  #org-info-js_console-label
    { font-size: 10px; font-weight: bold; white-space: nowrap; }
  .org-info-js_search-highlight
    { background-color: #ffff00; color: #000000; font-weight: bold; }
  .org-svg { width: 90%; }
  /*]]>*/-->
</style>
<link rel="stylesheet" type="text/css" href="https://fniessen.github.io/org-html-themes/src/readtheorg_theme/css/htmlize.css"/>
<link rel="stylesheet" type="text/css" href="https://fniessen.github.io/org-html-themes/src/readtheorg_theme/css/readtheorg.css"/>
<script src="https://ajax.googleapis.com/ajax/libs/jquery/2.1.3/jquery.min.js"></script>
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.4/js/bootstrap.min.js"></script>
<script type="text/javascript" src="https://fniessen.github.io/org-html-themes/src/lib/js/jquery.stickytableheaders.min.js"></script>
<script type="text/javascript" src="https://fniessen.github.io/org-html-themes/src/readtheorg_theme/js/readtheorg.js"></script>
<style> #content{max-width:1800px;}</style>
<script type="text/javascript">
/*
@licstart  The following is the entire license notice for the
JavaScript code in this tag.

Copyright (C) 2012-2020 Free Software Foundation, Inc.

The JavaScript code in this tag is free software: you can
redistribute it and/or modify it under the terms of the GNU
General Public License (GNU GPL) as published by the Free Software
Foundation, either version 3 of the License, or (at your option)
any later version.  The code is distributed WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.

As additional permission under GNU GPL version 3 section 7, you
may distribute non-source (e.g., minimized or compacted) forms of
that code without the copy of the GNU GPL normally required by
section 4, provided you include this license notice and a URL
through which recipients can access the Corresponding Source.


@licend  The above is the entire license notice
for the JavaScript code in this tag.
*/
<!--/*--><![CDATA[/*><!--*/
 function CodeHighlightOn(elem, id)
 {
   var target = document.getElementById(id);
   if(null != target) {
     elem.cacheClassElem = elem.className;
     elem.cacheClassTarget = target.className;
     target.className = "code-highlighted";
     elem.className   = "code-highlighted";
   }
 }
 function CodeHighlightOff(elem, id)
 {
   var target = document.getElementById(id);
   if(elem.cacheClassElem)
     elem.className = elem.cacheClassElem;
   if(elem.cacheClassTarget)
     target.className = elem.cacheClassTarget;
 }
/*]]>*///-->
</script>
<script type="text/x-mathjax-config">
    MathJax.Hub.Config({
        displayAlign: "center",
        displayIndent: "0em",

        "HTML-CSS": { scale: 100,
                        linebreaks: { automatic: "false" },
                        webFont: "TeX"
                       },
        SVG: {scale: 100,
              linebreaks: { automatic: "false" },
              font: "TeX"},
        NativeMML: {scale: 100},
        TeX: { equationNumbers: {autoNumber: "AMS"},
               MultLineWidth: "85%",
               TagSide: "right",
               TagIndent: ".8em"
             }
});
</script>
<script type="text/javascript"
        src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS_HTML"></script>
</head>
<body>
<div id="content">
<h1 class="title">Discordant variant analysis</h1>
<div id="table-of-contents">
<h2>Table of Contents</h2>
<div id="text-table-of-contents">
<ul>
<li><a href="#org656fa6a">1. Background</a></li>
<li><a href="#orgd936054">2. Aims</a></li>
<li><a href="#org8a9a6eb">3. Analysis pipeline</a></li>
<li><a href="#orgaf25115">4. BMI and T2D GWAS</a></li>
<li><a href="#orgfdf531e">5. Cross-referencing</a></li>
<li><a href="#org6885fc7">6. Assembly of concordant and discordant profiles</a></li>
<li><a href="#orgad85508">7. Phenome-wide scan - Data collection</a></li>
<li><a href="#orgf15e270">8. Phenome-wide comparison</a></li>
<li><a href="#org9a19a83">9. Phenome-wide comparison - Univariate analysis</a></li>
<li><a href="#org7f4033c">10. Results of univariate comparison</a></li>
<li><a href="#org77074f9">11. Phenome-wide comparison - SNP-Trait matrix</a></li>
<li><a href="#org514a3f4">12. PCA - a reminder</a></li>
<li><a href="#orgb4c3001">13. Random Forest (RF) - a reminder</a></li>
<li><a href="#org3b9aaa6">14. Clustering of traits using RF</a></li>
<li><a href="#org3635589">15. RF models</a></li>
<li><a href="#orgde61641">16. Clusters</a></li>
<li><a href="#org26bafd0">17. Trait selection - First stage</a></li>
<li><a href="#org5c170ed">18. Nested models</a></li>
<li><a href="#org375e58e">19. Trait selection - Final model</a></li>
<li><a href="#org9296480">20. External validation - BioVU PheWAS</a></li>
<li><a href="#org7a1f4f9">21. External validation - BioVU LabWAS</a></li>
<li><a href="#org61eff97">22. Mortality in UK Biobank - Concordant vs Discordant SNPs</a></li>
<li><a href="#org7f9f432">23. Mendelian Randomization</a></li>
<li><a href="#orge848ffc">24. Conclusions</a></li>
<li><a href="#org34a1a1d">25. Strengths and limitations</a></li>
<li><a href="#orgae55bab">26. Possible future applications</a></li>
</ul>
</div>
</div>

<div id="outline-container-org656fa6a" class="outline-2">
<h2 id="org656fa6a"><span class="section-number-2">1</span> Background</h2>
<div class="outline-text-2" id="text-1">
<ul class="org-ul">
<li>The relationship between obesity and cardiometabolic risk is highly heterogeneous.</li>
<li>Individuals in the same BMI level may display different cardiometabolic risk:
<ul class="org-ul">
<li>~ 30% of people with a BMI &gt; 30 kg/m<sup>2</sup> display a protective metabolic profile.</li>
<li>~ 30% of people with BMI &lt; 25 kg/m<sup>2</sup> develop conditions usually associated with obesity.</li>
</ul></li>
<li>Better risk stratification and more precise interventions may improve outcomes.</li>
<li>Obesity and T2D are closely linked.</li>
<li>T2D is in turn, linked to many cardiovascular diseases, and is life-threatening.</li>
<li>Factors uncoupling obesity from T2D risk may be used to define clinically relevant subgroups of obesity.</li>
</ul>
</div>
</div>
<div id="outline-container-orgd936054" class="outline-2">
<h2 id="orgd936054"><span class="section-number-2">2</span> Aims</h2>
<div class="outline-text-2" id="text-2">
<ul class="org-ul">
<li>Use genetics to stratify obesity into 2 subtypes:
<ul class="org-ul">
<li>Concordant (&uarr; BMI and &uarr; T2D)</li>
<li>Discordant (&uarr; BMI and &darr; T2D).</li>
</ul></li>
<li>Perform a phenome-wide scan to detect traits that reinforce this stratification.</li>
<li>Leverage this information for further stratification with potential clinical meaning.</li>
</ul>
</div>
</div>
<div id="outline-container-org8a9a6eb" class="outline-2">
<h2 id="org8a9a6eb"><span class="section-number-2">3</span> Analysis pipeline</h2>
<div class="outline-text-2" id="text-3">

<div class="figure">
<p><img src="./plots/aline_plot.png" alt="aline_plot.png" />
</p>
<p><span class="figure-number">Figure 1: </span>Analysis flowchart</p>
</div>
</div>
</div>
<div id="outline-container-orgaf25115" class="outline-2">
<h2 id="orgaf25115"><span class="section-number-2">4</span> BMI and T2D GWAS</h2>
<div class="outline-text-2" id="text-4">
<ul class="org-ul">
<li>We cross-referenced GWAS fo BMI (Yengo <i>et al.</i> [2018]) and T2D (Mahajan <i>et al</i> [2018]).</li>
<li>Inclusion/exclusion criteria:
<ul class="org-ul">
<li>Common biallelic variants (MAF &gt; 1%)</li>
<li>No INDELs</li>
<li>No potentially ambiguous palindromic SNPs (MAF &gt; 30%)</li>
<li>More than 20 % difference in MAF with 1000G EUR</li>
</ul></li>
<li>Clumpling (<i>r<sup>2</sup></i> &lt; 0.01 over 500kb window in 1000G EUR)</li>
</ul>
</div>
</div>
<div id="outline-container-orgfdf531e" class="outline-2">
<h2 id="orgfdf531e"><span class="section-number-2">5</span> Cross-referencing</h2>
<div class="outline-text-2" id="text-5">

<div class="figure">
<p><img src="./plots/gmirror.png" alt="gmirror.png" />
</p>
</div>
</div>
</div>
<div id="outline-container-org6885fc7" class="outline-2">
<h2 id="org6885fc7"><span class="section-number-2">6</span> Assembly of concordant and discordant profiles</h2>
<div class="outline-text-2" id="text-6">
<ul class="org-ul">
<li>We took the lead variants and aligned to the BMI increasing allele</li>
<li>They were stratified by their \(\beta\) coefficient for T2D.</li>
</ul>

<div class="figure">
<p><img src="./plots/profiles.png" alt="profiles.png" />
</p>
</div>
</div>
</div>
<div id="outline-container-orgad85508" class="outline-2">
<h2 id="orgad85508"><span class="section-number-2">7</span> Phenome-wide scan - Data collection</h2>
<div class="outline-text-2" id="text-7">
<ul class="org-ul">
<li>We queried the <a href="https://gwas.mrcieu.ac.uk/">MRC IEU GWAS database</a> for:
<ul class="org-ul">
<li>Associations of lead SNPs or nearest proxy if missing 
<ul class="org-ul">
<li><i>r<sup>2</sup></i> &lt; 0.01 over 500kb window in 1000G EUR</li>
</ul></li>
<li>Studies in EUR (congruency with the previous step</li>
<li>More than 500 individuals</li>
<li>More than 25 minor alleles in smallest group for binary traits</li>
<li>Studies with information for all reference SNPs</li>
<li>If multiple studies for a single trait:
<ul class="org-ul">
<li>The study with the highest sample size was selected</li>
</ul></li>
</ul></li>
<li>~ 3500 traits</li>
</ul>
</div>
</div>
<div id="outline-container-orgf15e270" class="outline-2">
<h2 id="orgf15e270"><span class="section-number-2">8</span> Phenome-wide comparison</h2>
<div class="outline-text-2" id="text-8">
<ul class="org-ul">
<li>We applied a two-stage analysis
<ul class="org-ul">
<li>Univariate comparison of profiles in each trait - Random effects meta-analysis</li>
<li>Clustering and selection of traits - Random forest</li>
</ul></li>
</ul>
</div>
</div>
<div id="outline-container-org9a19a83" class="outline-2">
<h2 id="org9a19a83"><span class="section-number-2">9</span> Phenome-wide comparison - Univariate analysis</h2>
<div class="outline-text-2" id="text-9">
<ul class="org-ul">
<li>Pooled concordant (\(\beta_C\)) and discordant (\(\beta_D\)) effects for each trait
<ol class="org-ol">
<li>Standardized beta coefficients: &nbsp; \(SE = 1/\sqrt{2 * MAF * (1 - MAF) * (n + Z^2)}\)</li>
<li>Random effects meta-analysis (Paule-Mandel \(\tau\) estimator)</li>
</ol></li>
<li>Difference between pooled estimates (<i>D</i>)</li>
</ul>
\begin{align*}
D & = |\beta_C - \beta_D|\\
\\
SE_D & = \sqrt{{SE_C}^2 + {SE_D}^2}\\
\\
Z & = \frac{D}{SE_D} \sim \mathcal{N}(0,\,{SE_D}^{2})
\end{align*}
<ul class="org-ul">
<li>Selection of traits: ( &nbsp; \(\beta_C \ne 0\) &nbsp; | &nbsp; \(\beta_D \ne 0\) &nbsp; ) &nbsp; &amp; &nbsp; \(D \ne 0\)
<ul class="org-ul">
<li>FDR 10%</li>
</ul></li>
<li>195 traits</li>
</ul>
</div>
</div>
<div id="outline-container-org7f4033c" class="outline-2">
<h2 id="org7f4033c"><span class="section-number-2">10</span> Results of univariate comparison</h2>
<div class="outline-text-2" id="text-10">

<div class="figure">
<p><img src="./plots/phewas_res.png" alt="phewas_res.png" />
</p>
<p><span class="figure-number">Figure 4: </span>Significant P values (FDR 10%) for difference between pooled estimates (D)</p>
</div>
</div>
</div>
<div id="outline-container-org77074f9" class="outline-2">
<h2 id="org77074f9"><span class="section-number-2">11</span> Phenome-wide comparison - SNP-Trait matrix</h2>
<div class="outline-text-2" id="text-11">
<ul class="org-ul">
<li>We filled a SNP-Trait matrix with \(z\)-scores in each trait:</li>
</ul>
\begin{align*}
z = \frac{\beta}{SE_{\beta}}
\end{align*}
<ul class="org-ul">
<li>Dimensions: 67 SNP x 195 traits
<ul class="org-ul">
<li>High dimensionality</li>
<li>Multicolinearity</li>
</ul></li>
<li>To tackle these problems we used the COVVSURF algorithm (Chavent <i>et al.</i> 2019), which combines:
<ul class="org-ul">
<li>Hierarchical agglomerative clustering - PCA</li>
<li>Random Forest</li>
</ul></li>
</ul>
</div>
</div>
<div id="outline-container-org514a3f4" class="outline-2">
<h2 id="org514a3f4"><span class="section-number-2">12</span> PCA - a reminder</h2>
<div class="outline-text-2" id="text-12">
<ul class="org-ul">
<li>PCs are linear combinations of original variables
<ul class="org-ul">
<li>Redundant variables can be summarized</li>
<li>The contribution of a variable in a PC = Squared loading (Pearson's <i>r<sup>2</sup></i>)</li>
</ul></li>
<li>Maximum possible information (variance) is summarized in PC1</li>
</ul>
</div>
</div>
<div id="outline-container-orgb4c3001" class="outline-2">
<h2 id="orgb4c3001"><span class="section-number-2">13</span> Random Forest (RF) - a reminder</h2>
<div class="outline-text-2" id="text-13">
<ul class="org-ul">
<li>RF is an ensemble of multiple decision trees</li>
<li>Can be used for classification or regression</li>
<li>Trees are grown using a random subset of columns and rows of the original data (in-bag)</li>
<li>Can be tested using out-of-bag (OOB) data</li>
<li>Averaging the OOB error rate gives an estimation of error rate / accuracy</li>
<li>The trees output probabilities (<i>votes</i>) to each datapoint assigned by trees</li>
<li>When testing, each variable is permuted out and the change in error rate is measured
<ul class="org-ul">
<li>If the error rate increase (i.e. importance score) is substantial then the variable is relevant</li>
</ul></li>
<li>Proximity matrix
<ul class="org-ul">
<li>N times two observations occupy the same terminal node in decision trees</li>
<li>Estimation of how similar observations are in the data space that is relevant for classification/regression</li>
<li>Can be used for clustering - Subgroups</li>
</ul></li>
</ul>
</div>
</div>
<div id="outline-container-org3b9aaa6" class="outline-2">
<h2 id="org3b9aaa6"><span class="section-number-2">14</span> Clustering of traits using RF</h2>
<div class="outline-text-2" id="text-14">
<ul class="org-ul">
<li>Dimensionality reduction into clusters of traits</li>
<li>Relevant for distinguishing between the two sets of SNPs</li>
<li>Non-parametric, data-driven</li>
<li>Process:
<ol class="org-ol">
<li>Agglomerative clustering using PCA results in:
<ul class="org-ul">
<li>Clustering tree with <i>k</i> = 1,2,3&#x2026;/p/ possible partitions</li>
<li>At every <i>k</i>: the 1PCs summarize each cluster</li>
</ul></li>
<li>Among every <i>k</i>: 1PC as predictors for RF</li>
<li>Model with minimum OOB error rate determines the optimal <i>k</i>.</li>
<li>Nested RF models
<ul class="org-ul">
<li>Starting with model with only the most important cluster</li>
<li>Ending with model including all clusters selected in <b>3</b></li>
<li>Final model - minimum OOB error rate</li>
</ul></li>
</ol></li>
</ul>
</div>
</div>
<div id="outline-container-org3635589" class="outline-2">
<h2 id="org3635589"><span class="section-number-2">15</span> RF models</h2>
<div class="outline-text-2" id="text-15">

<div class="figure">
<p><img src="./plots/kopt_phen.png" alt="kopt_phen.png" />
</p>
<p><span class="figure-number">Figure 5: </span>Error rate of RF models across every <i>k</i></p>
</div>
</div>
</div>
<div id="outline-container-orgde61641" class="outline-2">
<h2 id="orgde61641"><span class="section-number-2">16</span> Clusters</h2>
<div class="outline-text-2" id="text-16">

<div class="figure">
<p><img src="./plots/dend_phen.png" alt="dend_phen.png" />
</p>
<p><span class="figure-number">Figure 6: </span>Clustering tree with optimal partition</p>
</div>
</div>
</div>
<div id="outline-container-org26bafd0" class="outline-2">
<h2 id="org26bafd0"><span class="section-number-2">17</span> Trait selection - First stage</h2>
<div class="outline-text-2" id="text-17">

<div class="figure">
<p><img src="./plots/clusimp_phen.png" alt="clusimp_phen.png" />
</p>
<p><span class="figure-number">Figure 7: </span>Importance score of clusters</p>
</div>
</div>
</div>
<div id="outline-container-org5c170ed" class="outline-2">
<h2 id="org5c170ed"><span class="section-number-2">18</span> Nested models</h2>
<div class="outline-text-2" id="text-18">

<div class="figure">
<p><img src="./plots/nestedmodels.png" alt="nestedmodels.png" />
</p>
</div>
</div>
</div>
<div id="outline-container-org375e58e" class="outline-2">
<h2 id="org375e58e"><span class="section-number-2">19</span> Trait selection - Final model</h2>
<div class="outline-text-2" id="text-19">

<div class="figure">
<p><img src="./plots/phen_hm.png" alt="phen_hm.png" />
</p>
<p><span class="figure-number">Figure 9: </span>Importance score, squared loading, estimates and SNP clusters according to proximity</p>
</div>
</div>
</div>
<div id="outline-container-org9296480" class="outline-2">
<h2 id="org9296480"><span class="section-number-2">20</span> External validation - BioVU PheWAS</h2>
<div class="outline-text-2" id="text-20">

<div class="figure">
<p><img src="./plots/biovu_pw.png" alt="biovu_pw.png" />
</p>
<p><span class="figure-number">Figure 10: </span>Significant concordant/discordant effects in BioVU with significant differences (5% FDR) after exclusion of individuals with T1D</p>
</div>
</div>
</div>
<div id="outline-container-org7a1f4f9" class="outline-2">
<h2 id="org7a1f4f9"><span class="section-number-2">21</span> External validation - BioVU LabWAS</h2>
<div class="outline-text-2" id="text-21">

<div class="figure">
<p><img src="./plots/lw_biovu.png" alt="lw_biovu.png" />
</p>
<p><span class="figure-number">Figure 11: </span>Significant concordant/discordant effects in BioVU with significant differences (5% FDR) in Europeans and African Americans</p>
</div>
</div>
</div>
<div id="outline-container-org61eff97" class="outline-2">
<h2 id="org61eff97"><span class="section-number-2">22</span> Mortality in UK Biobank - Concordant vs Discordant SNPs</h2>
<div class="outline-text-2" id="text-22">

<div class="figure">
<p><img src="./plots/prs_surv.png" alt="prs_surv.png" />
</p>
<p><span class="figure-number">Figure 12: </span>Survival analysis of concordant/discordant profiles and subgroups</p>
</div>
</div>
</div>
<div id="outline-container-org7f9f432" class="outline-2">
<h2 id="org7f9f432"><span class="section-number-2">23</span> Mendelian Randomization</h2>
<div class="outline-text-2" id="text-23">

<div class="figure">
<p><img src="./plots/mr_res_all.png" alt="mr_res_all.png" />
</p>
</div>
</div>
</div>
<div id="outline-container-orge848ffc" class="outline-2">
<h2 id="orge848ffc"><span class="section-number-2">24</span> Conclusions</h2>
<div class="outline-text-2" id="text-24">
<ul class="org-ul">
<li>Concordant and discordant SNPs differ mainly in:
<ul class="org-ul">
<li>Liver enzymes - Central adiposity</li>
<li>Blood pressure</li>
<li>Lipids</li>
</ul></li>
<li>And differ less strongly in:
<ul class="org-ul">
<li>BMD</li>
<li>RBC counts</li>
<li>Overall adiposity</li>
</ul></li>
<li>The two subgroups are not homogeneous</li>
<li>The causal pathways might differ</li>
</ul>
</div>
</div>
<div id="outline-container-org34a1a1d" class="outline-2">
<h2 id="org34a1a1d"><span class="section-number-2">25</span> Strengths and limitations</h2>
<div class="outline-text-2" id="text-25">
<ul class="org-ul">
<li>Phenome-wide exploration for mechanisms uncoupling obesity from T2D</li>
<li>Less prone to reverse causality</li>
<li>Although the initial stage used only GWAS data for EUR, we saw differences in AA</li>
<li>Thresholds chosen affect:
<ul class="org-ul">
<li>Genetic factors chosen</li>
<li>Clusters built</li>
</ul></li>
</ul>
</div>
</div>
<div id="outline-container-orgae55bab" class="outline-2">
<h2 id="orgae55bab"><span class="section-number-2">26</span> Possible future applications</h2>
<div class="outline-text-2" id="text-26">
<ul class="org-ul">
<li>Other levels of the phenome</li>
<li>Random forest proximitry matrix to find more subtypes</li>
</ul>
</div>
</div>
</div>
<div id="postamble" class="status">
<p class="author">Author: Daniel Coral</p>
<p class="email">Email: <a href="mailto:daniel.coral@med.lu.se">daniel.coral@med.lu.se</a></p>
<p class="date">Created: 2021-06-10 Thu 11:24</p>
<p class="validation"><a href="http://validator.w3.org/check?uri=referer">Validate</a></p>
</div>
</body>
</html>
